File: Zipping and Unzipping files in .NET application/[VB]-Zipping and Unzipping files in .NET application/VB/ListViewItemSorters.vb
Imports System.Collections
Imports System.IO
Imports ComponentPro.Compression
Imports ComponentPro.IO

Namespace ArchiveManager
	Public Class TagInfo
		Public Name As String
		Public FullName As String
		Public Size As Long
		Public CompressedSize As Long
		Public CompressionRate As Double
		Public TypeName As String
		Public Modified As Date
		Public Crc As UInteger
		Public Attributes As FileAttributes

		Public Sub New(ByVal name As String, ByVal fullName As String, ByVal size As Long, ByVal compressedSize As Long, ByVal compressionRate As Double, ByVal typeName As String, ByVal modified As Date, ByVal crc As UInteger, ByVal attributes As FileAttributes)
			Me.Name = name
			Me.FullName = fullName
			Me.Size = size
			Me.CompressedSize = compressedSize
			Me.CompressionRate = compressionRate
			Me.TypeName = typeName
			Me.Modified = modified
			Me.Crc = crc
			Me.Attributes = attributes
		End Sub
	End Class

	''' <summary>
	''' Base comparer class for comparing two list view items.
	''' </summary>
	Public MustInherit Class ListViewItemComparerBase
		Implements IComparer
		Private ReadOnly _folderImageIndex As Integer
		Protected _sortOrder As SortOrder

		''' <summary>
		''' Initializes a new instance of the ListViewItemComparerBase.
		''' </summary>
		''' <param name="sortOrder">The sort order.</param>
		''' <param name="folderImageIndex">The defined folder image index.</param>
		''' <param name="folderLinkImageIndex">The defined folder link image index.</param>
		Protected Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer)
			_sortOrder = sortOrder

			_folderImageIndex = folderImageIndex
		End Sub

		Public Function Compare(ByVal xobject As Object, ByVal yobject As Object) As Integer Implements IComparer.Compare
			Dim x As ListViewItem = CType(xobject, ListViewItem)
			Dim y As ListViewItem = CType(yobject, ListViewItem)

			' Compare file to file or folder to folder only.
			Dim xIsFolder As Integer
			If x.ImageIndex = _folderImageIndex Then
				xIsFolder = 1
			Else
				xIsFolder = 0
			End If
			Dim yIsFolder As Integer
			If y.ImageIndex = _folderImageIndex Then
				yIsFolder = 1
			Else
				yIsFolder = 0
			End If

			If xIsFolder <> yIsFolder Then
				If _sortOrder = SortOrder.Ascending Then
					Return yIsFolder - xIsFolder
				Else
					Return (xIsFolder - yIsFolder)
				End If
			End If
			If _sortOrder = SortOrder.Ascending Then
				Return OnCompare(CType(x.Tag, TagInfo), CType(y.Tag, TagInfo), x.Text, y.Text)
			Else
				Return OnCompare(CType(y.Tag, TagInfo), CType(x.Tag, TagInfo), y.Text, x.Text)
			End If
		End Function

		Public MustOverride Function OnCompare(ByVal x As TagInfo, ByVal y As TagInfo, ByVal xtext As String, ByVal ytext As String) As Integer
	End Class

	''' <summary>
	''' Date time comparer.
	''' </summary>
	Public Class ListViewItemDateComparer
		Inherits ListViewItemComparerBase
		Public Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer)
			MyBase.New(sortOrder, folderImageIndex)
		End Sub

		Public Overrides Function OnCompare(ByVal x As TagInfo, ByVal y As TagInfo, ByVal xtext As String, ByVal ytext As String) As Integer
			Return Date.Compare(x.Modified, y.Modified)
		End Function
	End Class

	''' <summary>
	''' File attributes comparer.
	''' </summary>
	Public Class ListViewItemFileAttributesComparer
		Inherits ListViewItemComparerBase
		Public Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer)
			MyBase.New(sortOrder, folderImageIndex)
		End Sub

		Public Overrides Function OnCompare(ByVal x As TagInfo, ByVal y As TagInfo, ByVal xtext As String, ByVal ytext As String) As Integer
			Return x.Attributes - y.Attributes
		End Function
	End Class

	''' <summary>
	''' Name comparer.
	''' </summary>
	Public Class ListViewItemNameComparer
		Inherits ListViewItemComparerBase
		Public Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer)
			MyBase.New(sortOrder, folderImageIndex)
		End Sub

		Public Overrides Function OnCompare(ByVal x As TagInfo, ByVal y As TagInfo, ByVal xtext As String, ByVal ytext As String) As Integer
			Return String.Compare(xtext, ytext, True)
		End Function
	End Class

	''' <summary>
	''' Size comparer.
	''' </summary>
	Public Class ListViewItemSizeComparer
		Inherits ListViewItemComparerBase
		Public Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer)
			MyBase.New(sortOrder, folderImageIndex)
		End Sub

		Public Overrides Function OnCompare(ByVal x As TagInfo, ByVal y As TagInfo, ByVal xtext As String, ByVal ytext As String) As Integer
			If x.Size = y.Size Then
				Return 0
			End If
			If x.Size > y.Size Then
				Return 1
			Else
				Return -1
			End If
		End Function
	End Class

	''' <summary>
	''' Compressed Size comparer.
	''' </summary>
	Public Class ListViewItemCompressedSizeComparer
		Inherits ListViewItemComparerBase
		Public Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer)
			MyBase.New(sortOrder, folderImageIndex)
		End Sub

		Public Overrides Function OnCompare(ByVal x As TagInfo, ByVal y As TagInfo, ByVal xtext As String, ByVal ytext As String) As Integer
			If x.CompressedSize = y.CompressedSize Then
				Return 0
			End If
			If x.CompressedSize > y.CompressedSize Then
				Return 1
			Else
				Return -1
			End If
		End Function
	End Class

	''' <summary>
	''' Compressed Size comparer.
	''' </summary>
	Public Class ListViewItemTypeNameComparer
		Inherits ListViewItemComparerBase
		Public Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer)
			MyBase.New(sortOrder, folderImageIndex)
		End Sub

		Public Overrides Function OnCompare(ByVal x As TagInfo, ByVal y As TagInfo, ByVal xtext As String, ByVal ytext As String) As Integer
			Return String.Compare(x.TypeName, y.TypeName)
		End Function
	End Class

	''' <summary>
	''' Ratio comparer.
	''' </summary>
	Public Class ListViewItemRatioComparer
		Inherits ListViewItemComparerBase
		Public Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer)
			MyBase.New(sortOrder, folderImageIndex)
		End Sub

		Public Overrides Function OnCompare(ByVal x As TagInfo, ByVal y As TagInfo, ByVal xtext As String, ByVal ytext As String) As Integer
			If x.CompressionRate = y.CompressionRate Then
				Return 0
			End If

			If x.CompressionRate > y.CompressionRate Then
				Return 1
			Else
				Return -1
			End If
		End Function
	End Class

	''' <summary>
	''' Checksum comparer.
	''' </summary>
	Public Class ListViewItemChecksumComparer
		Inherits ListViewItemComparerBase
		Public Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer)
			MyBase.New(sortOrder, folderImageIndex)
		End Sub

		Public Overrides Function OnCompare(ByVal x As TagInfo, ByVal y As TagInfo, ByVal xtext As String, ByVal ytext As String) As Integer
			Return CInt(x.Crc - y.Crc)
		End Function
	End Class
End Namespace

File: Zipping and Unzipping files in .NET application/[C#]-Zipping and Unzipping files in .NET application/C#/ListViewItemSorters.cs
using System;
using System.Collections;
using System.IO;
using System.Windows.Forms;
using ComponentPro.Compression;
using ComponentPro.IO;

namespace ArchiveManager
{
    public class TagInfo
    {
        public string Name;
        public string FullName;
        public long Size;
        public long CompressedSize;
        public double CompressionRate;
        public string TypeName;
        public DateTime Modified;
        public uint Crc;
        public FileAttributes Attributes;

        public TagInfo(string name, string fullName, long size, long compressedSize, double compressionRate, string typeName, DateTime modified, uint crc, FileAttributes attributes)
        {
            Name = name;
            FullName = fullName;
            Size = size;
            CompressedSize = compressedSize;
            CompressionRate = compressionRate;
            TypeName = typeName;
            Modified = modified;
            Crc = crc;
            Attributes = attributes;
        }
    }
    
    /// <summary>
    /// Base comparer class for comparing two list view items.
    /// </summary>
    public abstract class ListViewItemComparerBase : IComparer
    {
        private readonly int _folderImageIndex;
        protected SortOrder _sortOrder;

        /// <summary>
        /// Initializes a new instance of the ListViewItemComparerBase.
        /// </summary>
        /// <param name="sortOrder">The sort order.</param>
        /// <param name="folderImageIndex">The defined folder image index.</param>
        /// <param name="folderLinkImageIndex">The defined folder link image index.</param>
        protected ListViewItemComparerBase(SortOrder sortOrder, int folderImageIndex)
        {
            _sortOrder = sortOrder;

            _folderImageIndex = folderImageIndex;
        }

        public int Compare(object xobject, object yobject)
        {
            ListViewItem x = (ListViewItem) xobject;
            ListViewItem y = (ListViewItem) yobject;

            // Compare file to file or folder to folder only.
            int xIsFolder = (x.ImageIndex == _folderImageIndex) ? 1 : 0;
            int yIsFolder = (y.ImageIndex == _folderImageIndex) ? 1 : 0;

            if (xIsFolder != yIsFolder)
                return _sortOrder == SortOrder.Ascending ? (yIsFolder - xIsFolder) : (xIsFolder - yIsFolder);
            return _sortOrder == SortOrder.Ascending ? OnCompare((TagInfo)x.Tag, (TagInfo)y.Tag, x.Text, y.Text) : OnCompare((TagInfo)y.Tag, (TagInfo)x.Tag, y.Text, x.Text);
        }

        public abstract int OnCompare(TagInfo x, TagInfo y, string xtext, string ytext);
    }

    /// <summary>
    /// Date time comparer.
    /// </summary>
    public class ListViewItemDateComparer : ListViewItemComparerBase
    {
        public ListViewItemDateComparer(SortOrder sortOrder, int folderImageIndex) : base(sortOrder, folderImageIndex)
        {
        }

        public override int OnCompare(TagInfo x, TagInfo y, string xtext, string ytext)
        {
            return DateTime.Compare(x.Modified, y.Modified);
        }
    }

    /// <summary>
    /// File attributes comparer.
    /// </summary>
    public class ListViewItemFileAttributesComparer : ListViewItemComparerBase
    {
        public ListViewItemFileAttributesComparer(SortOrder sortOrder, int folderImageIndex)
            : base(sortOrder, folderImageIndex)
        {
        }

        public override int OnCompare(TagInfo x, TagInfo y, string xtext, string ytext)
        {
            return x.Attributes - y.Attributes;
        }
    }

    /// <summary>
    /// Name comparer.
    /// </summary>
    public class ListViewItemNameComparer : ListViewItemComparerBase
    {
        public ListViewItemNameComparer(SortOrder sortOrder, int folderImageIndex)
            : base(sortOrder, folderImageIndex)
        {
        }

        public override int OnCompare(TagInfo x, TagInfo y, string xtext, string ytext)
        {
            return string.Compare(xtext, ytext, true);
        }
    }

    /// <summary>
    /// Size comparer.
    /// </summary>
    public class ListViewItemSizeComparer : ListViewItemComparerBase
    {
        public ListViewItemSizeComparer(SortOrder sortOrder, int folderImageIndex)
            : base(sortOrder, folderImageIndex)
        {
        }

        public override int OnCompare(TagInfo x, TagInfo y, string xtext, string ytext)
        {
            if (x.Size == y.Size)
                return 0;
            return x.Size > y.Size ? 1 : -1;
        }
    }

    /// <summary>
    /// Compressed Size comparer.
    /// </summary>
    public class ListViewItemCompressedSizeComparer : ListViewItemComparerBase
    {
        public ListViewItemCompressedSizeComparer(SortOrder sortOrder, int folderImageIndex)
            : base(sortOrder, folderImageIndex)
        {
        }

        public override int OnCompare(TagInfo x, TagInfo y, string xtext, string ytext)
        {
            if (x.CompressedSize == y.CompressedSize)
                return 0;
            return x.CompressedSize > y.CompressedSize ? 1 : -1;
        }
    }

    /// <summary>
    /// Compressed Size comparer.
    /// </summary>
    public class ListViewItemTypeNameComparer : ListViewItemComparerBase
    {
        public ListViewItemTypeNameComparer(SortOrder sortOrder, int folderImageIndex)
            : base(sortOrder, folderImageIndex)
        {
        }

        public override int OnCompare(TagInfo x, TagInfo y, string xtext, string ytext)
        {
            return string.Compare(x.TypeName, y.TypeName);
        }
    }

    /// <summary>
    /// Ratio comparer.
    /// </summary>
    public class ListViewItemRatioComparer : ListViewItemComparerBase
    {
        public ListViewItemRatioComparer(SortOrder sortOrder, int folderImageIndex)
            : base(sortOrder, folderImageIndex)
        {
        }

        public override int OnCompare(TagInfo x, TagInfo y, string xtext, string ytext)
        {
            if (x.CompressionRate == y.CompressionRate)
                return 0;

            return x.CompressionRate > y.CompressionRate ? 1 : -1;
        }
    }

    /// <summary>
    /// Checksum comparer.
    /// </summary>
    public class ListViewItemChecksumComparer : ListViewItemComparerBase
    {
        public ListViewItemChecksumComparer(SortOrder sortOrder, int folderImageIndex)
            : base(sortOrder, folderImageIndex)
        {
        }

        public override int OnCompare(TagInfo x, TagInfo y, string xtext, string ytext)
        {
            return (int)(x.Crc - y.Crc);
        }
    }
}
